src/core/service.c

```diff
@@ -1782,14 +1781,34 @@ static void service_enter_start_post(Service *s) {
         service_enter_stop(s, SERVICE_FAILURE_RESOURCES);
 }
 
-static void service_kill_control_processes(Service *s) {
-        char *p;
+static int service_kill_control_processes(Service *s) {
+        int r;
 
-        if (!UNIT(s)->cgroup_path)
-                return;
+        assert(s);
+
+        if (s->control_pid > 0) {
+                r = kill_and_sigcont(s->control_pid, SIGKILL);
+                if (r < 0)
+                        log_unit_debug_errno(UNIT(s), r, "Failed to kill control process, ignoring: %m");
```

Comment:
It'd be nice to print the PID here (and possibly also comm, but maybe that's too much). Just to make any debugging easier (and NFS-related process might be unkillable, so this is not entirely theoretical). "SIGKILL" should be mentioned too.